lib/ansible/modules/web_infrastructure/acme_certificate.py

```diff
@@ -261,7 +263,10 @@
       type: string
       sample: .well-known/acme-challenge/evaGxfADs6pSRb2LAv9IZf17Dt3juxGJ-PCt92wr-oA
     resource_value:
-      description: the value the resource has to produce for the validation
+      description:
+        - the value the resource has to produce for the validation
+        - "I(Note:) The value for C(tls-alpn-01) challenges is a Base64 encoded string.
+           It must be Base64-decoded before it can be used."
```

Comment:
I am not sure if I understand this. Must the value be base64 encoded by the users? like using the filter `{{ resource_value | b64encode }}`?

Reply:
Hmm, yes, it's not formulated very clearly, it seems :) The data needed for the challenge is a binary string. Since when trying this out, I ran into some trouble, so I decided to return the base64-encoded result. So when you want to use it, you probably have to do `{{ resource_value | b64decode }}`. If that works well, I'm not sure how well Jinja2 templating goes with binary data.

The main problem with an example here is that there's essentially nothing in real world which can handle this yet ;-) But I can try to come up with something.

Anyway, do you know (or have some examples where this is done) how to handle binary data with Ansible variables / templating? If that's possible, I'd like to get rid of base64 encoding here.

Reply:
Actually, I tried adding support for the ACME validation X509 extension to `openssl_csr` first; adding it works fine, but trying to check if is in an existing CSR might not be possible because the interface of pyOpenSSL is so horrible (at least not with some amount of monkey-patching and/or very ugly hacks to try to access inner data-structures).

Reply:
I just tried removing the base64 encoding, and then using `{{ resource_value | b64encode }}` to pass the data on to the validation satisfier script (which still expects base64-encoded data), and it doesn't work. Looks like the binary data gets destroyed somewhere between the `acme_certificate` module and the URI module which passes the data on. (Also, the debug output becomes pretty ugly due to the binary data in it.)

Reply:
Ok. I apparently modified my local testing container and didn't rebuilt it after changing the satisfyer back. After rebuilding it, both the original base64 version succeeded, as well as the non-base64 encoded version. So it *does* work without base64 encoding. Just debug output looks ugly.
I'll remove base64 encoding then.

Reply:
I've re-introduced base64 encoding; see below. The explanation is now hopefully better...